src/Featurizers/Components/FrequencyEstimator.h

```diff
@@ -0,0 +1,167 @@
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Data
+    // |
+    // ----------------------------------------------------------------------
+    static constexpr char const * const     NameValue = FrequencyEstimatorName;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    FrequencyTrainingOnlyPolicy(void);
+
+    void fit(InputType const &input);
+    FrequencyAnnotation complete_training(void);
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Types
+    // |
+    // ----------------------------------------------------------------------
+    using FrequencyType                     = std::chrono::system_clock::duration;
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Data
+    // |
+    // ----------------------------------------------------------------------
+    FrequencyType _frequency;
+    InputType _lastObserved;
+
+    bool _hasSeenValues;
+};
+}
+
+/////////////////////////////////////////////////////////////////////////
+///  \typedef       FrequencyEstimator
+///  \brief         A training-only class that finds the min, max
+///
+template <
+    size_t MaxNumTrainingItemsV=std::numeric_limits<size_t>::max()
+>
+using FrequencyEstimator                       = TrainingOnlyEstimatorImpl<Details::FrequencyTrainingOnlyPolicy, MaxNumTrainingItemsV>;
+
+
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// |
+// |  Implementation
+// |
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+// ----------------------------------------------------------------------
+
+// ----------------------------------------------------------------------
+// |
+// |  FrequencyAnnotation
+// |
+// ----------------------------------------------------------------------
+inline FrequencyAnnotation::FrequencyAnnotation(FrequencyAnnotation::FrequencyType value) :
+    Value(std::move(value)) {
+}
+
+// ----------------------------------------------------------------------
+// |
+// |  Details::FrequencyTrainingOnlyPolicy
+// |
+// ----------------------------------------------------------------------
+Details::FrequencyTrainingOnlyPolicy::FrequencyTrainingOnlyPolicy(void) :
+    _frequency(std::chrono::system_clock::duration::max().count()),
+    _hasSeenValues(false) {
+}
+
+void Details::FrequencyTrainingOnlyPolicy::fit(InputType const &input) {
+    if (!_hasSeenValues) {

```

Comment:
Switch to ntd::optional.
Refactor to only one return statement.
Add a const reference after the value has been set.